### ISSUE TITLE
불량 이미지 feature 를 통한 데이터 중복 (Imbalance)문제 해결

### ISSUE THREAD

**Issue by Hyeonjha** [Hyeonjha/AIV_ImageEmbedding#10]
Jul 9

Original comment by JihoonPark93

main branch 에 data-gatter 폴더에 이미지 가져오는 부분 코드 수정해서 올려뒀어요

코드 실행하시면 downloaded 폴더 안에 class 별로 이미지들이 모여있을텐데 class 폴더들 중 가장 이미지 개수가 "적은" 개수를 기준으로 각 불량 클래스별로 이미지를 모아주는 task 를 진행하시면 됩니다

- 예를 들어, TIP class 가 30장이고 나머지 불량들이 전부 30장 이상씩 있다면 나머지 불량 유형들을 가장 unique 한 30장만 뽑아두고 유사한 이미지들은 클래스 폴더 하위에 similar_image 라는 경로로 모아주시면 돼요!
이미지는 저도 똑같이 받을 수 있으니까 이미지는 깃에 안올리고 작업해주셔도 될 것 같습니다!

꼭 embedding feature 만 사용할 필요는 없고, 이미지 크기나 가로세로 비율 같은 정보를 함께 사용하시는 게 좋을 것 같아요!
embedding 의 경우 resize 를 해서 처리하다 보니 geometry 정보가 사라질 수 있어서 해당 정보들을 같이 쓰는 게 좋을 것 같습니다

**Comment by Hyeonjha** [comment]
Jul 16

어제 진행할때 unique 이미지 분류 갯수가 들쭉날쭉해서 오늘 이 부분 수정하였고 관련 코드 Issue11_Imbalance 폴더에 업로드 했습니다!
추가해야할 부분이나 수정해야 할 부분이 있으면 말씀해주세요:)

+) Issue9에서 10만개 넣고 돌리기가 꽤 오래 걸리네요..! 지금 weaviate에 5만개 정도 저장이 돼서 오늘 집에서 계속 돌아가게해두고 성능 평가 하도록 하겠습니다!

**Comment by Hyeonjha** [comment]
Jul 16

Original comment by JihoonPark93

unique image 를 계산하는 부분 로직을 수정해야 할 것 같습니다!
현재는 image list 를 순회하면서 처음 min_count 보다 개수가 적을 땐 무조건 unique_images list 에 추가되고, 이후 유사도를 판단해서 list 에 추가하는 식으로 동작하는 것 같은데요
동일한 불량 클래스 이미지들 안에서 feature vector 유사도가 "가장 낮은" feature 이미지를 뽑도록 해야할 것 같아요! unique_list 에 포함된 features 와 가장 유사도가 떨어진 이미지를 찾도록 확인 부탁드려요!

**Comment by Hyeonjha** [comment]
Jul 18

Issue11_Imbalance/result.ipynb 파일 내용

![IMG_C786C2224008-1](https://github.com/user-attachments/assets/e6417781-3259-48e0-8abb-ea5629ddffaa)

전체 data set (총 2470개) 
- 시간이 오래 걸려서 데이터 수가 많은 DUST(10614→ 507)와 SCRATCH(3623 → 504)는 줄여서 사용. 
- 나머지 클래스들은 모두 그대로 사용. 

test set (총 111개) 
- 각 클래스별 10개의 이미지, RING은 1개 

unique data만 사용했을때 그 이미지들이 어떻게 보면 이상치만 사용한거니까 성능이 안좋게 나오는 것 같아서 k-means clustering으로 class 대표할 수 있는 이미지 추가하여 성능 평가해봤습니다!
성능이 오르긴 해서 더 다뤄보고 싶은데 k-means가 아니라 그냥 간단하게 유사도가 높은 순서로 min_count만큼 추가하여 사용해보려고 하는데 괜찮을까요!?
추가 다른 방법도 생각해보겠습니다:)

**Comment by Hyeonjha** [comment]
Jul 18

**[ Pretrained Weight ]**

               전체 data   unique(+size) unique RING x(+size)  유사도 HL(+ size)   유사도 HL RING x(+ size)   

    ConvNeXt       63       41 (26)        54 (48)                49 (41)             68           
    ResNet18       57       32 (21)        50 (43)                49 (42)             61           
    ResNet50       59       35 (29)        55 (43)                50 (44)             62           

**[ Random Weight ]**
 
               전체 data   unique(+size) unique RING x(+size)  유사도 HL(+ size)   유사도 HL RING x(+ size)   

    ConvNeXt       36       26 (22)        30 (29)                32 (28)             40           
    ResNet18       57       33 (24)        55 (48)                46 (41)             57           
    ResNet50       59       33 (27)        45 (45)                52 (41)             59           


|| |
|:---|---|
|전체 data|        전체 Train data 사용 |        
|unique|          RING 기준 유니크 이미지 분류하여 훈련 (각 클래스별 5개의 이미지)      | 
|unique RING x|   RING 제외하여 유니크 이미지 분류하여 훈련 (DOT 기준 - 각 클래스별 42개의 이미지. RING은 5개)       |
|유사도 HL|         RING 기준 유사도 높은(similar) 이미지 + 유사도 낮은(unique) 이미지 분류 (각 클래스별 10개의 이미지)   |   
|유사도 HL RING x|  RING 제외하여 유사도 높고 낮은 이미지 분류 (DOT 기준 - 각 클래스별 84개의 이미지. RING은 5개) |
|(+size)|         size_info 포함하여 분류한 unique 이미지로 모델 훈련      |

**Comment by Hyeonjha** [comment]
Jul 19

지훈님!
classification model(efficientnet) 기준으로 학습시키는 코드 push 해놨습니다.
(Issue11_Imbalance/testModelRandom.py)
근데 제일 작은 train data set으로 학습시키는데도 꽤 오래 걸리네요,,!
혹시 잘못된 부분 있는지 한번만 확인 부탁드립니다!
다음주에 이어서 진행하도록 하겠습니다:)

**Comment by Hyeonjha** [comment]
Jul 19

Original comment by JihoonPark93


classification 학습할 때 epoch 을 충분히 늘려서 해봐야 할 것 같아요!
아 GPU 디바이스가 없어서 엄청 느리겠네요ㅠ 담주에 서버 남는 게 있는지 확인해보고 남는 서버에서 테스트하는 게 좋을 것 같네요!
